crates/red_knot_test/src/lib.rs

```diff
@@ -324,15 +316,15 @@ struct TestFile {
     backtick_offsets: Vec<BacktickOffsets>,
 }
 
-fn create_diagnostic_snapshot<D: OldDiagnosticTrait>(
+fn create_diagnostic_snapshot(
     db: &mut db::Db,
     relative_fixture_path: &Utf8Path,
     test: &parser::MarkdownTest,
-    diagnostics: impl IntoIterator<Item = D>,
+    diagnostics: impl IntoIterator<Item = Diagnostic>,
 ) -> String {
     let display_config = DisplayDiagnosticConfig::default().color(false);
 
-    let mut snapshot = String::new();
+    let mut snapshot: Vec<u8> = Vec::new();
```

Comment:
I believe you mentioned in one of your reviews that it was a very intentional decision to implement `print` over `std::io::Write`. I don't remember the details but I noticed that it results in many places where it's necessary to use `Vec<u8>` in combination with a `String::from_utf8().unwrap` which is somewhat unfortunate. 

I don't really have a solution for this, it's just something I noticed (e.g. also in my tests).

Reply:
Yeah... right... good point. The reason was related to the linear type fakery. Namely, using `std::io::Write` as a destination makes it "more likely" that printing is actually going to stdout or similar, and not just some string in memory that could get accidentally ignored.

Reply:
I don't think I have any strong feelings about reverting this back to just a `Display` impl or some such (the internal renderer already works that way, so it's an overall small change). One note though is that the `unwrap()` calls you're referencing I believe only occur in tests.

Reply:
I'm leaning towards offering a `Display` impl. But might do that in a follow-up PR.

Reply:
> One note though is that the unwrap() calls you're referencing I believe only occur in tests.

This was wrong! See: https://github.com/astral-sh/ruff/pull/17130#discussion_r2024815794